src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1610 @@
+        &mut self,
+        hash: &'desc sha256d::Hash,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if let Some(StackElement::Push(preimage)) = self.pop() {
+            if preimage.len() != 32 {
+                return Some(Err(Error::HashPreimageLenghtMismatch))
+            }
+            if sha256d::Hash::hash(preimage) == *hash {
+                self.push(StackElement::Satisfied);
+                Some(Ok(SatisfiedConstraint::HashLock {
+                    hash: HashLockType::Hash256(hash),
+                    preimage,
+                }))
+            } else {
+                self.push(StackElement::Dissatisfied);
+                None
+            }
+        } else {
+            Some(Err(Error::UnexpectedStackEnd))
+        }
+    }
+
+    fn evaluate_hash160<'desc, Pk, Pkh>(
+        &mut self,
+        hash: &'desc hash160::Hash,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if let Some(StackElement::Push(preimage)) = self.pop() {
+            if preimage.len() != 32 {
+                return Some(Err(Error::HashPreimageLenghtMismatch))
+            }
+            if hash160::Hash::hash(preimage) == *hash {
+                self.push(StackElement::Satisfied);
+                Some(Ok(SatisfiedConstraint::HashLock {
+                    hash: HashLockType::Hash160(hash),
+                    preimage,
+                }))
+            } else {
+                self.push(StackElement::Dissatisfied);
+                None
+            }
+        } else {
+            Some(Err(Error::UnexpectedStackEnd))
+        }
+    }
+
+    fn evaluate_ripemd160<'desc, Pk, Pkh>(
+        &mut self,
+        hash: &'desc ripemd160::Hash,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if let Some(StackElement::Push(preimage)) = self.pop() {
+            if preimage.len() != 32 {
+                return Some(Err(Error::HashPreimageLenghtMismatch))
+            }
+            if ripemd160::Hash::hash(preimage) == *hash {
+                self.push(StackElement::Satisfied);
+                Some(Ok(SatisfiedConstraint::HashLock {
+                    hash: HashLockType::Ripemd160(hash),
+                    preimage,
+                }))
+            } else {
+                self.push(StackElement::Dissatisfied);
+                None
+            }
+        } else {
+            Some(Err(Error::UnexpectedStackEnd))
+        }
+    }
+    /// Helper function to evaluate a checkmultisig which takes the top of the stack as input
+    /// signature and validates it in order of pubkeys. If the first signature
+    /// is satisfied by second public keys, other signatures are not checked
+    /// against the first pubkey.
+    ///
```

Comment:
I don't understand this comment

Reply:
Reclarified it. Hopefully, it's more clear now.